[Team-20 / BE - Dong] 숙소 조회,  좋아요를 누른 숙소 기능

moto6 opened the pull request:
# intro
안녕하세요  Team-20 BE  Dong 입니다

지금 많은 부분이 부족한 코드인데 조금은 마음이 싱숭생숭해서 부족한 코드지만PR 날립니다 😥😥😥😥😥

# Done
- 숙소 정보 리스트 
- 좋아요 누른 숙소만 리턴해주는 기능
- 메인페이지에서 추천여행지, 테마숙소 리턴해주는 기능

# Todo(앞으로 진행할것들)
- 배포 자동화
- 테스트코드

# 궁금한점
- 입력받는 부분에서 검색조건을 받아서 조건에 맞는 숙소를 돌려줘야하는데, 데이터베이스 where절로 처리해야할지, 자바코드로  if 문을 활용해야 할지 고민입니다
- recursive query 기능을 이용해서 "서울시" 를 검색하면, 서울시의 하위 행정구역인 "서초구", "강남구" 등을 검색할수 있는 쿼리문을 구현했는데, 이를 프로젝트에 실제로 어떻게 적용해야할지 모르겠습니다.

moto6 committed:
first commit

malaheaven committed:
feat: Dto 멤버 내용 추가

moto6 committed:
imges upload

moto6 committed:
Merge pull request #12 from d-h-k/BE/feature/7/dto

[BE] Dto 멤버 내용 추가

moto6 committed:
Merge pull request #13 from d-h-k/BE/feature/8/img

imges upload

malaheaven committed:
feat: Dao 샘플을 구현

moto6 committed:
웹페이지 아이콘 추가

moto6 committed:
오어스 기능 구현을 위한 백엔드 간이 페이지 구성

moto6 committed:
oauth 기능 구현을 위한 GithubAccessTokenResponse 객체 추가

필드변수로    private String accessToken;
          private String tokenType;
          private String scope;
가 있음

moto6 committed:
oauth 기능 구현을 위한 부수적인 것들

moto6 committed:
세션 사용을 위한 그레이들 셋팅

moto6 committed:
Update README.md

moto6 committed:
github oauth 를 통해 jwt 토큰 얻어오는데 성공

moto6 committed:
메인페이지에 팝업 해 줄 정보들을 백엔드에서 보내주도록 파일 생성

- 클래스 및 생성자와 구현을 위한 메서드 껍데기만 만듬
- 데이터가 포함된 로직은 아직 구현 안됨

moto6 committed:
메인페이지에 팝업 해 줄 정보들을 백엔드에서 보내주도록 로직과 데이터 추가

moto6 committed:
메인페이지에 팝업 해 줄 정보들을 백엔드에서 보내주도록 로직과 데이터 추가

moto6 committed:
Merge pull request #18 from d-h-k/BE/feature/14/daoSample

feat: Dao 샘플을 구현

moto6 committed:
Merge pull request #22 from d-h-k/team-20

Update README.md

moto6 committed:
Merge branch 'dev-BE' into BE/feature/15/oauth

moto6 committed:
Merge pull request #23 from d-h-k/BE/feature/15/oauth

Be/feature/15/oauth

moto6 committed:
Merge pull request #24 from d-h-k/BE/feature/16/suggestions

Be/feature/16/suggestions

moto6 committed:
숙소 관련 JDBC Templete 필드변수 추가

moto6 committed:
숙소 평점 저장을 위해 Double 형으로 저장

moto6 committed:
숙소 목업 API (/airbnb) 제공을 위한  수정

- 생성자 수정
- 컨트롤러에 getSample() 메서드 추가
- AccommodationDTO 필드변수 도메인과 동일하게 수정

moto6 committed:
숙소검색 API에서 모든 데이터가 전달되지 않는 문제 해결

moto6 committed:
데이터 7곳의 숙소 정보가 리턴되도록 다섯개의 데이터 셋 추가

moto6 committed:
숙소 이미지 데이터셋 추가

moto6 committed:
숙소 이미지 데이터셋이 나오도록 변경

moto6 committed:
숙소 이미지 데이터셋이 나오도록 수정

moto6 committed:
그레이들 컴파일 에러 해결을 위해서 11버전 >> 8로 변경

moto6 committed:
배포 스크립트 추가

moto6 committed:
(iOS)위시리스트 누른 숙소 리스트 돌려주기 #45

위 이슈 대응을 위해 코드 추가

moto6 committed:
branch protection test

moto6 committed:
기본적인 테스트 코드작성

moto6 committed:
숙소 검색결과 리스트를 리턴해줄때 검색쿼리스트링을 포함해(에코잉으로) 보내주도록 기능 수정

moto6 committed:
숙소 검색할때 검색조건을 쿼리스트링으로 받는다.

랜더링사이드의 편의를 위해서 API의 모든 파라미터는 있거나 없거나 모든 상황에 대응되도록 제작

moto6 committed:
불필요한 메서드 정리
- 테스트용 코드이므로 테스트파일로 이동예정(after 테스트코드 브랜치 통합 이후)

moto6 committed:
Merge pull request #47 from d-h-k/BE/feature/40/query_echo

Be/feature/40/query echo

moto6 committed:
Merge pull request #48 from d-h-k/BE/feature/33/testcode

기본적인 테스트 코드작성

moto6 committed:
Merge branch 'dev-BE' into BE/feature/45/likeAcmdOnly

moto6 committed:
Merge pull request #49 from d-h-k/BE/feature/45/likeAcmdOnly

(iOS)위시리스트 누른 숙소 리스트 돌려주기 #45

moto6 committed:
불필요한 메서드 정리
- 테스트용 코드이므로 테스트파일로 이동예정(after 테스트코드 브랜치 통합 이후)

moto6 committed:
파라미터가 하나뿐인 생성자 추가

moto6 committed:
데이터베이스 테스트코드 추가 및 정리

moto6 committed:
sql update

moto6 committed:
프론트에서 쿼리스트링으로 체크인 체크아웃 날짜데이터를 문자열로 받았을때, 이를 LocalDate로 변환해주는 코드

moto6 committed:
문자열에서 날짜로 변경할때 에러가 있는경우 런타임 익셉션을 발동시켜주는 클래스

moto6 committed:
기존 도메인 폴더에 있던 해당 클래스를 유틸클래스로 옮김

moto6 committed:
에어비엔비 프로젝트 기능 구현을 위한 Guest, Reservation 도메인 도입

moto6 committed:
의존 클래스 위치 변경 반영

moto6 committed:
불필요한 주석 삭제

moto6 committed:
불필요한 주석 삭제

moto6 committed:
Merge pull request #60 from d-h-k/BE/feature/26/recursive_query_for_search_location

sql update

moto6 committed:
지역명 검색 리커시브 쿼리

moto6 committed:
사용가능한 숙소 리스트에서 검색조건에 맞는 숙소를 반영할수 있는 코드 추가

moto6 committed:
백엔드 목표관리용 마크다운파일 추가

moto6 committed:
코드 리포매팅

moto6 committed:
불필요한 주석 삭제

moto6 committed:
코드 정리/리포매팅

moto6 committed:
CORS 문제 대응을 위한 코드 삽입

moto6 committed:
불필요한 개행 삭제

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/auth/AuthController.java:
```diff
@@ -0,0 +1,92 @@
+    private final String ISSUER = "";
+    private final String CLIENT_ID = "your client id";
+    private final String CLIENT_SECRET = "your client secret";
+    private final String CODE = "";
```
컨트롤러에 위치해야 할 정보들이 아닌 것 같습니다. 어떻게 분리할 수 있을까요?

moto6 replied:
제가 너무 안일한 생각을 했습니다. 

아주아주 간단한 모듈(auth)이 라서, 서비스계층과 컨트롤러 계층을 나눌 생각을 하지 않았는데, 말씀을 듣고 나니 컨트롤러 계층에서 가지고 있기 부적절한 정보가 맞습니다. 수정 완료했습니다!

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/auth/AuthController.java:
```diff
@@ -0,0 +1,92 @@
+        GithubAccessTokenResponse accessToken = getAccessToken(code, gitHubRequest)
+                .orElseThrow(() -> new RuntimeException("바디 없음"));
+
+        System.out.println(accessToken);
+        System.out.println(gitHubRequest);
```
안됩니다.

moto6 replied:
넵 수정했습니다!

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/auth/AuthController.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+
+    private Optional<User> getUserFromGitHub(GithubAccessTokenResponse accessToken, RestTemplate gitHubRequest) {
+        RequestEntity<Void> request = RequestEntity
+                .get(GITHUB_USER_URI)
+                .header("Accept", "application/json")
+                .header("Authorization", "token " + accessToken.getAccessToken())
+                .build();
+
+        ResponseEntity<User> response = gitHubRequest
+                .exchange(request, User.class);
+
+        return Optional.ofNullable(response.getBody());
+    }
+
+    private Optional<GithubAccessTokenResponse> getAccessToken(String code, RestTemplate gitHubRequest) {
+        RequestEntity<GithubAccessTokenRequest> request = RequestEntity
+                .post(GITHUB_ACCESS_TOKEN_URI)
+                .header("Accept", "application/json")
+                .body(new GithubAccessTokenRequest(CLIENT_ID, CLIENT_SECRET, code));
+
+        ResponseEntity<GithubAccessTokenResponse> response = gitHubRequest
+                .exchange(request, GithubAccessTokenResponse.class);
+
+        return Optional.ofNullable(response.getBody());
+    }
```
별도의 서비스나 클라이언트 클래스(컴포넌트 클래스) 가 구성돼야 하겠지요?
컨트롤러에 있으면 안될 녀석들인 것 같은데 어떻게 생각하시는지 궁금합니다.

moto6 replied:
네 위에서 말씀해주셔서 해당 부분은 코드에 반영했습니다

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/auth/AuthController.java:
```diff
@@ -0,0 +1,92 @@
+                    .sign(algorithm);
+        } catch (JWTCreationException exception) {
+            //Invalid Signing configuration / Couldn't convert Claims.
+            throw new RuntimeException(exception);
```
`JWTCreationException` 이라는 라이브러리에서 제공하는 checked exception을 그냥 `RuntimeException` 으로 던지네요?
좋은 예외처리일까요? 어떻게 생각하시나요? `RuntimeException` 으로 충분한가요?

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/auth/AuthController.java:
```diff
@@ -0,0 +1,92 @@
+                    .withClaim("login", user.getLogin())
+                    .withClaim("name", user.getName())
+                    .withIssuer(ISSUER)
+                    .sign(algorithm);
```
마찬가지로 별도의 컴포넌트 구성돼야 하겠죠?

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/auth/Jwt.java:
```diff
@@ -0,0 +1,13 @@
+package com.example.airbnb.auth;
+
+public class Jwt {
```
너무 포괄적인 클래스명이고 잠재적으로 다른 라이브러리가 제공하는 API나 클래스명과 중첩될 위험성도 커보이네요.
어떻게 잘 수정할 수 있으려나요?

moto6 replied:
생각해보니까, 토큰 클래스에서 별도로 뭔가 처리하는 로직도 없고, 단순히  String Class 를 한번 Wrapping한 클래스이고, 불필요하다 생각해서 해당 클래스를 삭제하였습니다. 

처음에는 뭔가 JWT 클래스를 만들어두고, JWT 토큰을 복호화해서 정보를 가져올 필요가 있을꺼같아 클래스를 따로 만들어줬는데, 막상 제작해보니, 이미 User관련 정보를 모두 들고있어 불필요하다 생각이 들었습니다. 해당부분 더티코드인데 발견해주셔서 감사합니다!

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/controller/AccommodationController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @GetMapping
+    public AccommodationListDTO getSample(@RequestParam(required = false) String location,
+                                          @RequestParam(required = false) String checkin,
+                                          @RequestParam(required = false) String checkout,
+                                          @RequestParam(required = false) Integer adults,
+                                          @RequestParam(required = false) Integer children,
+                                          @RequestParam(required = false) Integer infants) {
```
[이 링크](https://stackoverflow.com/questions/16942193/spring-mvc-complex-object-as-get-requestparam) 한 번 참고해볼까요?

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/controller/SuggestionController.java:
```diff
@@ -0,0 +1,18 @@
+
+    @GetMapping
+    public MainPageDTO mainPageSuggestions() {
+        return SuggestionsService.mainPage();
```
서비스 클래스를 static 호출하네요....괜찮을까요? 
`SuggestionsService` 는 Bean으로 관리되고 있나요? 그렇다면 왜 주입받지 않는 거죠?
의존성 주입의 이점을 포기하면서까지 이런 방향으로 구현해야 할 이유가 있었는지 궁금합니다.

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/domain/Accommodation.java:
```diff
@@ -0,0 +1,121 @@
+        return this;
+    }
+
+    //String badge;
```
모든 게터에 이런 주석이 붙어있는데 그렇게 의미있는 주석은 아닌 듯 보이기는 합니다.

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/dto/LocationDTO.java:
```diff
@@ -0,0 +1,4 @@
+package com.example.airbnb.dto;
+
+public class LocationDTO {
+}
```
음....

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/exception/JWTCreationException.java:
```diff
@@ -0,0 +1,9 @@
+package com.example.airbnb.exception;
+
+public class JWTCreationException extends RuntimeException {
+    final static String MESSAGE = "error";
```
우선 이 상수를 이 클래스에서 관리해야겠다고 생각한 이유가 궁금합니다.
그리고 `MESSAGE` 라는 변수명이 어떠한 의미를 갖는지가 궁금하군요.

moto6 replied:
앗 해당부분은 제가 놓쳤네요. 조금 더 명확한 문자열의 이름과 메시지를 작성하였습니다!

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/exception/UnExpectedTimeFormat.java:
```diff
@@ -0,0 +1,9 @@
+package com.example.airbnb.exception;
+
+public class UnExpectedTimeFormat extends RuntimeException {
+    private static final String UNEXPECTEDTIMEFORMAT = "this Time-string-Format does not support this applications";
```
언더스코어를 적절히 넣어야겠죠.
`UNEXPECTED_TIME_FORMAT` 으로 되겠네요.

moto6 replied:
넵 해당부분 수정하였습니다!

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/repository/AccommodationDAO.java:
```diff
@@ -0,0 +1,84 @@
+
+
+    private JdbcTemplate jdbcTemplate;
+    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;
```
이 두 컴포넌트 다 주입받을 수 있을 겁니다. 
Spring Context에서 어떤 컴포넌트를 기본으로 빈 등록해서 주입해줄 수 있는지 알아보시는 것도 큰 도움이 되겠네요.

moto6 replied:
넵 리뷰해주신 부분 동의합니다! 

- NamedParameterJdbcTemplate 은 사용하지 않아서 해당부분은 걷어냈습니다.
- Spring Context에서 어떤 컴포넌트를 기본으로 빈 등록해서 주입해줄 수 있는지 학습한 내용은 아래 링크에 정리해뒀습니다. 
  - https://velog.io/@d-h-k/Spring-Context에서-주입해줄-수-있는-빈Bean-컴포넌트
- 좀 많이 부족한데 차근차근 채워나가겠습니다!

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/service/AccommodationService.java:
```diff
@@ -0,0 +1,93 @@
+        return accommodationList;
+    }
+
+    private boolean isPeopleCountValidated(SearchConditions conditions) {
+        if (0 >= conditions.getAdults()) {
+            return false;
+        }
+        if (0 >= conditions.getChildren()) {
+            return false;
+        }
+        return 0 < conditions.getInfants();
+    }
```
`SearchConditions` 로 이동시키면 어떨까요?

moto6 replied:
넵 해당부분 코드 수정했습니다!

wheejuni commented on BE/airbnb/src/main/java/com/example/airbnb/service/SuggestionsService.java:
```diff
@@ -0,0 +1,35 @@
+    private static final List<RecommendedDTO> recommendedDTOList = new ArrayList<>();
+    private static final List<ThemeDTO> themeDTOList = new ArrayList<>();
+
+    static {
```
DB에 미리 넣어두는 방식은 안되려나요?

moto6 replied:
음 지금은 해당 Recommended, Theme 정보가 추가되거나 삭제되지 않는 형태라서 아직 DB 는 아니지만, 장기적인 관점에서는 확장성을 위해서 DB 에 넣어주는 방식이 좋을꺼같습니다!

wheejuni reviewed (changes requested):
데이터베이스에 질의하는 로직을 if로 처리하면, 나중에 데이터가 방대해져서 페이징 등이 발생할 때 어떻게 대처할지도 잘 고민해 보셔야 할 겁니다. 개인적으로 추천하는 방식은 아닙니다. 

"행정구역" 에 관한 개념이 추상화되어 있으면 되겠죠? 서로간의 포함관계가 존재하면 될 겁니다. 

그런데 지금 그런 부분보다도, "관성적으로 코딩하지 않는 방법" 에 대한 고민이 필요한 시점이라고 판단됩니다.
그런 고민이 부족하신 듯 보여서 조금 아쉽다는 느낌은 들었습니다. 여튼 수고 많으셨고 수정해야 할 부분에 대해 다시 살펴보셨으면 좋겠습니다. 💯

moto6 committed:
Merge branch 'dev-BE' of https://github.com/d-h-k/airbnb into dev-BE

moto6 committed:
변수명 가독성을 올리기 위해 적절한 언더스코어 활용

moto6 committed:
options 필드 추가

moto6 committed:
DAO ->> Repository 이름 변경

moto6 committed:
Option 필드 데이터 추가

moto6 committed:
Option 필드 데이터 추가

moto6 committed:
Accommodation 필드 에러 수정
- reviewCount가 정상적으로 실행되지 않고 null 값이 나오던 문제를 수정
- Options 필드 추가

moto6 committed:
Auth 인증 관련 프론트엔드 협조요청사항 반영

moto6 committed:
잘못된 데이터 수정

moto6 committed:
FE 요청- location추가  issue #87

- fe 측 요청으로 location 관련한 컬럼 데이터 추가

moto6 committed:
fe측과 oauth 연동 테스트

moto6 committed:
FE 요청대응 ) auth 정보 리턴시 DTO를 만들어서 사용자 정보 추가해서 보내줌

moto6 committed:
숙소 정보를 추가하는 API 추가 >>미완성 상태 1단계

moto6 committed:
프론트엔드 요청으로 깃헙 프로필사진 정보도 같이 오도록 수정

moto6 committed:
코드 리포매팅

moto6 committed:
깃허브에서 유저 프로필 사진을 가져오도록 기능을 수정

moto6 committed:
좋아요 기능 추가

moto6 committed:
클라이언트 시크릿 코드 삭제

moto6 committed:
Auth에 컨트롤러와 서비스 계층이 분리되지 않아서 부적절한 정보들이 컨트롤러 계층에 위치해 있던 문제 해결

- 리뷰사항 https://github.com/codesquad-members-2021/airbnb/pull/53#discussion_r642483047
- 코드 클리닝을 진행하고 서비스계층과 컨트롤러 계층의 역할을 구분함

moto6 committed:
- 컴포넌트 주입을 필요한 부분에서만 받도록 수정

moto6 committed:
불필요한 주석 삭제

moto6 committed:
숙소 검색 인원수 체크 로직을 기존에 서비스에서 SearchConditions 클래스로 이동

- https://github.com/codesquad-members-2021/airbnb/pull/53#discussion_r642496195
- 위 리뷰 사항 참고함

moto6 committed:
런타임 익셉션에서 메시지 정보가 부정확하던 것을 좀더 구체적이고 명확하게 수정함

- 위 리뷰 참고
- https://github.com/codesquad-members-2021/airbnb/pull/53#discussion_r642491668

moto6 committed:
불필요한  JWT 클래스제거

- 단순한 String class의 래핑클래스이므로 삭제
